src/NServiceBus.Transport.RabbitMQ/Receiving/MessagePump.cs

```diff
@@ -251,6 +251,7 @@ async Task Process(BasicDeliverEventArgs message)
                     catch (Exception ex)
                     {
                         ++numberOfDeliveryAttempts;
+                        headers = messageConverter.RetrieveHeaders(message);
                         var errorContext = new ErrorContext(ex, headers, messageId, message.Body ?? new byte[0], transportTransaction, numberOfDeliveryAttempts);
                         errorHandled = await onError(errorContext).ConfigureAwait(false) == ErrorHandleResult.Handled;
```

Comment:
Do we also need to protect against potential modification from the `onError` call?

Reply:
if somebody uses a custom recoverability policy, this could potentially be an issue. I'm not sure our own recoverability logic could leak anything but I'll double check.

Reply:
by default, there are no header modifications going on in the error handling logic for immediate retries. However, if you write your custom recoverability policy. Also, code subscribing to the `MessageToBeRetried` event in the event aggregator has full access to the context, including headers. This event is only available internally in core at the moment (there were ideas to make them public though). Thoughts?

Reply:
Would I be correct in thinking that the only way to guard against this would be to read the headers immediately before the call to `onMessage(messageContext)`, as well as before every call to `onError(errorContext)` as we are doing here?

Reply:
@adamralph but core is only accepting `Dictionary<string, string>` so there is no elegant way to work around that I think?

Reply:
@timbussmann I guess we could derive from `Dictionary` and instiate that type in the header retrieval. We'd pass that object to `onMessage` and `onError`. It could track whether any of the values for existing keys are reassigned, or whether any keys are added or removed. Back in the transport code we could ask the object to give us the original dictionary, which would be just itself if unmodified or a new object otherwise. Although I suspect it may be difficult to write the internals of this new type without resulting in more, often unecessary, allocations on the hotpath. 🤷‍♂️

Reply:
It would make things messier, but we could move the initial `RetrieveHeaders` and `RetrieveMessageId` code inside the while loop, right before the `onMessage` call.

That way we always get a new header collection, but we aren't allocating an additional collection in the case of a message that is processed on the first try.

Reply:
@bording yeah, that's what I meant up here: https://github.com/Particular/NServiceBus.RabbitMQ/pull/482#discussion_r209936999

It means paying that cost for every call to `onMessage` and `onError`, but perhaps it's a necessary evil.

Reply:
What cost? How would that be different from how it currently is?